File: community/cypher/cypher-compiler-2.1/src/main/scala/org/neo4j/cypher/internal/compiler/v2_1/ast/Token.scala
/**
 * Copyright (c) 2002-2014 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.compiler.v2_1.ast

import org.neo4j.cypher.internal.compiler.v2_1._
import org.neo4j.cypher.internal.compiler.v2_1.LabelId
import org.neo4j.cypher.internal.compiler.v2_1.ast.Identifier
import org.neo4j.cypher.internal.compiler.v2_1.RelTypeId

trait Token[T <: TokenId] extends ASTNode with SymbolicName {
  def id: Option[T]
}

case class LabelToken(name: String, id: Option[LabelId] = None)(val position: InputPosition) extends Token[LabelId]
case class PropertyKeyToken(name: String, id: Option[PropertyKeyId] = None)(val position: InputPosition) extends Token[PropertyKeyId]
case class RelTypeToken(name: String, id: Option[RelTypeId] = None)(val position: InputPosition) extends Token[RelTypeId]

object LabelToken {
  def fromIdentifier(identifier: Identifier): LabelToken = LabelToken(identifier.name)(identifier.position)
}

object PropertyKeyToken {
  def fromIdentifier(identifier: Identifier): PropertyKeyToken = PropertyKeyToken(identifier.name)(identifier.position)
}

object RelTypeToken {
  def fromIdentifier(identifier: Identifier): RelTypeToken = RelTypeToken(identifier.name)(identifier.position)
}

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/Token.java
/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.internal.kernel.api;

import org.neo4j.internal.kernel.api.exceptions.KernelException;

/**
 * Token creation and lookup.
 */
public interface Token
{
    /**
     * Value indicating the a token does not exist in the graph.
     */
    int NO_TOKEN = -1;

    /**
     * Returns a label id for a label name. If the label doesn't exist prior to
     * this call it gets created.
     */
    int labelGetOrCreateForName( String labelName ) throws KernelException;

    /**
     * Returns a property key id for a property key. If the key doesn't exist prior to
     * this call it gets created.
     */
    int propertyKeyGetOrCreateForName( String propertyKeyName ) throws KernelException;

    int relationshipTypeGetOrCreateForName( String relationshipTypeName ) throws KernelException;

    void labelCreateForName( String labelName, int id ) throws KernelException;

    void propertyKeyCreateForName( String propertyKeyName, int id ) throws KernelException;

    void relationshipTypeCreateForName( String relationshipTypeName, int id ) throws KernelException;

    /**
     * Return the id of the provided label, or NO_TOKEN if the label isn't known to the graph.
     *
     * @param name The label name.
     * @return the label id, or NO_TOKEN
     */
    int nodeLabel( String name );

    /**
     * Return the id of the provided relationship type, or NO_TOKEN if the type isn't known to the graph.
     *
     * @param name The relationship type name.
     * @return the relationship type id, or NO_TOKEN
     */
    int relationshipType( String name );

    /**
     * Return the id of the provided property key, or NO_TOKEN if the property isn't known to the graph.
     *
     * @param name The property key name.
     * @return the property key id, or NO_TOKEN
     */
    int propertyKey( String name );
}
